python_modules/dagster-graphql/dagster_graphql/implementation/execution/backfill.py

```diff
@@ -318,3 +324,76 @@ def resume_partition_backfill(
 
     graphene_info.context.instance.update_backfill(backfill.with_status(BulkActionStatus.REQUESTED))
     return GrapheneResumeBackfillSuccess(backfill_id=backfill_id)
+
+
+def retry_partition_backfill(
+    graphene_info: "ResolveInfo", backfill_id: str
+) -> "GrapheneRetryBackfillSuccess":
+    from ...schema.backfill import GrapheneRetryBackfillSuccess
+
+    backfill = graphene_info.context.instance.get_backfill(backfill_id)
+    if not backfill:
+        check.failed(f"No backfill found for id: {backfill_id}")
+
+    if backfill.status not in BULK_ACTION_COMPLETED_STATUSES:
+        raise DagsterInvariantViolationError(
+            f"Cannot retry backfill {backfill_id} because it is still in progress."
+        )
+
+    if backfill.is_asset_backfill:
+        asset_backfill_data = backfill.get_asset_backfill_data(graphene_info.context.asset_graph)
+        if (
+            asset_backfill_data.failed_and_downstream_subset.num_partitions_and_non_partitioned_assets
+            == 0
+        ):
+            raise DagsterInvariantViolationError(
+                "Cannot retry an asset backfill that has no failed assets."
+            )
+        asset_graph = graphene_info.context.asset_graph
+        assert_permission_for_asset_graph(
+            graphene_info,
+            asset_graph,
+            list(asset_backfill_data.failed_and_downstream_subset.asset_keys),
+            Permissions.LAUNCH_PARTITION_BACKFILL,
+        )
+
+        new_backfill = PartitionBackfill.from_asset_graph_subset(
+            backfill_id=make_new_backfill_id(),
+            asset_graph_subset=asset_backfill_data.failed_and_downstream_subset,
```

Comment:
might need to make the retried asset graph subset from the targeted subset - the materialized subset instead. Need to look into the failure conditions of the backfill daemon some more. If the daemon could mark a backfill as failed without moving all unmaterialized partitions into the failed subset, then need to make sure those are included in the retry